app/src/main/java/de/westnordost/streetcomplete/data/maptiles/MapTilesDownloader.kt

```diff
@@ -30,19 +31,20 @@ class MapTilesDownloader @Inject constructor(
         val time = currentTimeMillis()
 
         coroutineScope {
-            for (tile in getDownloadTileSequence(bbox)) {
-                launch {
-                    val result = try {
-                        downloadTile(tile.zoom, tile.x, tile.y)
-                    } catch (e: Exception) {
-                        DownloadFailure
-                    }
-                    ++tileCount
-                    when (result) {
-                        is DownloadFailure -> ++failureCount
-                        is DownloadSuccess -> {
-                            if (result.alreadyCached) cachedSize += result.size
-                            else downloadedSize += result.size
```

Comment:
So, there's actually a concurrency bug here: `++` and `+=` are not atomic, so the different coroutines we `launch` here can overwrite each other's values mid-flight. It's the same [shared mutable state problem mentioned in the docs](https://kotlinlang.org/docs/shared-mutable-state-and-concurrency.html#the-problem). This doesn't happen too often, because we're not downloading thousands of tiles, but it does happen. In a single test of around-75-percent-cached tiles, I got
```
06-19 20:10:40.084  3899  4031 I MapTilesDownload: Caught 163 tiles of 164
```
<details>
<summary>Code</summary>

```diff
diff --git a/app/src/main/java/de/westnordost/streetcomplete/data/maptiles/MapTilesDownloader.kt b/app/src/main/java/de/westnordost/streetcomplete/data/maptiles/MapTilesDownloader.kt
index 4ff149169..da95426a2 100644
--- a/app/src/main/java/de/westnordost/streetcomplete/data/maptiles/MapTilesDownloader.kt
+++ b/app/src/main/java/de/westnordost/streetcomplete/data/maptiles/MapTilesDownloader.kt
@@ -12,6 +12,7 @@ import okhttp3.*
 import okhttp3.internal.Version
 import java.io.IOException
 import java.lang.System.currentTimeMillis
+import java.util.concurrent.atomic.AtomicInteger
 import javax.inject.Inject
 import kotlin.coroutines.resume
 
@@ -29,6 +30,7 @@ class MapTilesDownloader @Inject constructor(
         var downloadedSize = 0
         var cachedSize = 0
         val time = currentTimeMillis()
+        val reliableCounter = AtomicInteger()
 
         coroutineScope {
             listOf(
@@ -39,6 +41,7 @@ class MapTilesDownloader @Inject constructor(
                     launch {
                         val result = downloadTile(source, tile.zoom, tile.x, tile.y)
                         ++tileCount
+                        reliableCounter.incrementAndGet()
                         when (result) {
                             is DownloadFailure -> ++failureCount
                             is DownloadSuccess -> {
@@ -52,6 +55,7 @@ class MapTilesDownloader @Inject constructor(
         }
         val seconds = (currentTimeMillis() - time) / 1000.0
         val failureText = if (failureCount > 0) ". $failureCount tiles failed to download" else ""
+        Log.i(TAG, "Caught $tileCount tiles of ${reliableCounter.get()}")
         Log.i(TAG, "Downloaded $tileCount tiles (${downloadedSize / 1000}kB downloaded, ${cachedSize / 1000}kB already cached) in ${seconds.format(1)}s$failureText")
     }
```
</details>

Since the data here are simple, the quickest solution would be to change out the integers for AtomicIntegers (and, I may end up doing this anyway for performance). However, I want to clean up this part of the code anyway, so I'm planning on (read: mostly done with) converting it to use [flows](https://kotlinlang.org/docs/flow.html), which are basically "sequences but for coroutines".

Reply:
But it doesn't matter if they overwrite each other. Whether the log reads that either 163 or 164 tiles have been downloaded hardly matters. I'd prefer not to use `AtomicInteger` because that's Java and I try to not have the core classes depend on Java dependencies.

Reply:
Well, it means that the `downloadedSize` is not accurate; it is a little larger than the reported value.

Reply:
I *would* like to fix this before proceeding, because having an accurate picture of download sizes is important for deciding whether to always download aerial images or only when the setting is enabled.

The idea I'm currently planning to go with is to iterate *indexed* inside a coroutineScope and store the DownloadStats result into an ArrayList by index. Then I'll sum the stats at the end, outside the coroutineScope, where I know all the coroutines have finished.

Reply:
It makes the code more makes though. Before you commit this,  I'd like to ask you to log via both methods ans see by how much the two numbers differ. I would expect that the imprecision through race conditions is miniscule and this doesn't matter.  But i may be wrong.